### PR TITLE
MAINT Clean up leftover ``Array = Any`` aliases in jax/_src/**.py

### PR DESCRIPTION
**Pull request by superbobry**
This seemingly minor change uncovered a few inconsistencies in existing code. I silenced the new type errors via ``type: ignore` and added a TODO to investigate later.

### CODE REVIEW

**Comment by jakevdp** [comment]
If we're going to land this, I think we should fix rather than silence the errors as part of this PR.

**Comment by superbobry** [comment]
I fixed two out of three errors. 

The remaining one is tricky, because it overlaps dynamic shapes and ragged tensors and is not documented enough for me to figure out what the "right" (intended) types are.

File: jax/_src/interpreters/batching.py [hunk 1]
```diff
@@ -148,8 +148,10 @@ def _sorted_ragged_axis(stacked_axis, ragged_axes):
   return RaggedAxis(stacked_axis, tuple(sorted(ragged_axes, key=lambda p: p[0])))
 
 def make_batch_axis(
-    ndim: int, stacked_axis: int, ragged_axes: list[tuple[int, Array]]
-  ) -> int | RaggedAxis:
+    ndim: int,
+    stacked_axis: int,
+    ragged_axes: list[tuple[int, Array | core.Var | core.Tracer]],
```
Comment (jakevdp) [inline jax/_src/interpreters/batching.py:153]: `core.Tracer` shouldn't be necessary here, because it is a subclass of `Array`

File: jax/_src/lax/convolution.py [hunk 2]
```diff
@@ -343,7 +337,7 @@ def conv_transpose(lhs: Array, rhs: Array, strides: Sequence[int],
   if transpose_kernel:
     # flip spatial dims and swap input / output channel axes
     rhs = _flip_axes(rhs, np.array(dn.rhs_spec)[2:])
-    rhs = np.swapaxes(rhs, dn.rhs_spec[0], dn.rhs_spec[1])
+    rhs = np.swapaxes(rhs, dn.rhs_spec[0], dn.rhs_spec[1])  # type: ignore[assignment]
```
Comment (jakevdp) [inline jax/_src/lax/convolution.py:340]: Better fix would be to replace this with `rhs.swapaxes(dn.rhs_spec[0], dn.rhs_spec[1])` rather than relying on the undocumented ability of`np.swapaxes` to dispatch to `jax.numpy`.

File: jax/_src/nn/functions.py [hunk 3]
```diff
@@ -165,7 +165,7 @@ def log_sigmoid(x: Array) -> Array:
   return -softplus(-x)
 
 @jax.jit
-def elu(x: Array, alpha: Array = 1.0) -> Array:
```
Comment (jakevdp) [inline jax/_src/nn/functions.py:168]: Array inputs in this module should be annotated `ArrayLike`, otherwise we're likely to have spurious errors if someone calls these functions on e.g. a numpy array or scalar

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/interpreters/batching.py (hunk 1 above)
Comment (superbobry) [reply jax/_src/interpreters/batching.py:153]: Done.

File: jax/_src/lax/convolution.py (hunk 2 above)
Comment (superbobry) [reply jax/_src/lax/convolution.py:340]: Nice! I haven't thought of using a method.

File: jax/_src/nn/functions.py (hunk 3 above)
Comment (superbobry) [reply jax/_src/nn/functions.py:168]: Done, PTAL.

File: jax/_src/nn/functions.py [hunk 4]
```diff
@@ -186,11 +187,12 @@ def elu(x: Array, alpha: Array = 1.0) -> Array:
   See also:
     :func:`selu`
   """
-  safe_x = jnp.where(x > 0, 0., x)
-  return jnp.where(x > 0, x, alpha * jnp.expm1(safe_x))
+  x = jnp.asarray(x)
+  safe_x = jnp.where(jnp.greater(x, 0), 0., x)
+  return jnp.where(jnp.greater(x, 0), x, alpha * jnp.expm1(safe_x))
```
Comment (jakevdp) [inline jax/_src/nn/functions.py:192]: I'd probably go back to using `>` instead of `jnp.greater` – the original code was easier to read

File: jax/_src/nn/functions.py [hunk 5]
```diff
@@ -353,9 +357,11 @@ def glu(x: Array, axis: int = -1) -> Array:
   See also:
     :func:`sigmoid`
   """
-  size = x.shape[axis]
```
Comment (jakevdp) [inline jax/_src/nn/functions.py:356]: you can always do `np.shape(x)[axis]` if you don't want to explicitly cast to array

Comment (superbobry) [reply jax/_src/nn/functions.py:356]: Ack. I decided to keep the original version and convert `x` to an array instead.

File: jax/_src/nn/functions.py [hunk 6]
```diff
@@ -317,20 +329,22 @@ def gelu(x: Array, approximate: bool = True) -> Array:
     x : input array
     approximate: whether to use the approximate or exact formulation.
   """
+  numpy_util.check_arraylike("gelu", x)
 
   # Promote to nearest float-like dtype.
-  x = x.astype(dtypes.to_inexact_dtype(x.dtype))
+  x_arr = jnp.asarray(x)
+  x_arr = x_arr.astype(dtypes.to_inexact_dtype(x_arr.dtype))
```
Comment (jakevdp) [inline jax/_src/nn/functions.py:336]: You could replace these three lines with `x_arr = numpy_util.promote_args_inexact("gelu", x)`

File: jax/_src/nn/functions.py [hunk 7]
```diff
@@ -431,17 +447,17 @@ def softmax(x: Array,
     :func:`log_softmax`
   """
   if jax.config.jax_softmax_custom_jvp:
-    return _softmax(x, axis, where, initial)
+    return _softmax(x, axis, where, initial)  # type: ignore[return-value]
```
Comment (jakevdp) [inline jax/_src/nn/functions.py:450]: Do you understand why this ignore statement is necessary? It looks like `_softmax` is annotated to return `Array`

**Review by jakevdp — APPROVED** [review]
Looks good! Another couple small comments.

File: jax/_src/nn/functions.py (hunk 6 above)
Comment (superbobry) [reply jax/_src/nn/functions.py:336]: Nice, done!

File: jax/_src/nn/functions.py (hunk 7 above)
Comment (superbobry) [reply jax/_src/nn/functions.py:450]: Not entirely, but it seem mypy is confused by the `@functools.partial` decorator in `_softmax`

```
jax/_src/nn/functions.py:450: error: Incompatible return value type (got "ReturnValue", expected "Array")  [return-value]
```

`ReturnValue` in the error message is a type parameter of `jax.custom_jvp`, which mypy just keeps as is instead of substituting for the return type of `_softmax` (or for `Any`):

https://github.com/google/jax/blob/2d068a1caa97ebde662604bb95298cff2abb0afa/jax/_src/custom_derivatives.py#L234

Comment (jakevdp) [reply jax/_src/nn/functions.py:453]: OK, fair enough. The inner workings of mypy are sometimes mysterious...

**Review by jakevdp — CHANGES_REQUESTED** [review]
Last thing before merging – can you please squash your changes into a single commit?

**Review by jakevdp — APPROVED** [review]
